text/0000-contract-metadata.md

```diff
@@ -0,0 +1,306 @@
+- Proposal Name: `contract_metadata`
+- Start Date: 2019-06-25
+- NEP PR: [nearprotocol/neps#0003](https://github.com/nearprotocol/NEPs/pull/3)
+
+# Summary
+[summary]: #summary
+
+This NEP introduces contract metadata, which summarizes the content of a given contract in json format.
+Contract metadata allows developers to easily list the methods in a contract and potentially display them to users to
+provide transparency. 
+It also provides the ability to validate input data in frontend thanks to the signatures of contract methods provided
+by the contract metadata.
+Contract metadata is similar to contract abi in some other blockchains such as Ethereum or Substrate but is more geared
+towards readability and developer friendliness.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there is no convenient way for a developer to programmatically list the methods of a contract that is deployed on chain
+because the contract code stored on chain is the compiled wasm code, where the parameters and type information are already mangled (even though
+method name is not, it is still cumbersome to extract them from wasm binary).
+As a result, if developers want to display the methods of a contract to an user of the app to provide transparency,
+especially in the case of financial apps, they have no choice but to hardcode them in the front end, which is suboptimal in many ways.
+Furthermore, if developers want to get the list of methods in some contract for some downstream task like data analysis, or interacting
+with other contracts, they have no way of doing so.
+In addition, because of the lack of signature of the contract functions, there is no way that the frontend can validate
+that input data is of the right format, which might lead to some spamming attacks.
+Contract metadata aims to solve the aforementioned problems by providing a convenient way 
+for developers to list the methods of contracts.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+With contract metadata, developers, instead of annotating view and change methods in the `initContract` function in `main.js`,
+will instead annotate the methods of a contract by decorators. We propose that view functions be annotated with the `@view` decorator
```

Comment:
```suggestion
will instead annotate the methods of a contract by decorators. We propose that view functions be annotated with the `@view` decorator in AssemblyScript (in Rust there is no need for additional decorators)
```

Reply:
How do we plan to do this in Rust? This decorator doesn't enforces anything but rather serves as a way to document the function.

Reply:
Technically it's distinguished by the borrowing type of ABI method in implementation (`fn method(&mut self)` or `method(&self)` for view). See
https://github.com/nearprotocol/near-bindgen/blob/master/examples/status-message/src/lib.rs